torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,354 @@
+                                                    torch.Tensor) else tensor
+            return tensor
+
+        attr = self._attr_cls.cast(attr)
+        if isinstance(attr.index,
+                      slice) and (attr.index.start, attr.index.stop,
+                                  attr.index.step) == (None, None, None):
+            attr.index = None
+
+        return to_type(self._get_tensor(attr))
+
+    @abstractmethod
+    def _remove_tensor(self, attr: TensorAttr) -> bool:
+        r"""Implemented by :obj:`FeatureStore` subclasses."""
+        pass
+
+    def remove_tensor(self, attr: TensorAttr) -> bool:
+        r"""Removes a :obj:`FeatureTensorType` object from the feature store.
+
+        Args:
+            attr (TensorAttr): any relevant tensor attributes that correspond
+                to the tensor to remove. See :obj:`TensorAttr` documentation
+                for required and optional attributes. It is the job of
+                implementations of a FeatureStore to store this metadata in a
+                meaningful way that allows for tensor deletion from a
+                :obj:`TensorAttr` object.
+
+        Returns:
+            bool: whether deletion was succesful.
+        """
+        attr = self._attr_cls.cast(attr)
+        self._remove_tensor(attr)
+
+    def update_tensor(self, tensor: FeatureTensorType,
+                      attr: TensorAttr) -> bool:
+        r"""Updates a :obj:`FeatureTensorType` object with a new value.
+        implementor classes can choose to define more efficient update methods;
+        the default performs a removal and insertion.
+
+        Args:
+            tensor (FeatureTensorType): the features to be added.
+            attr (TensorAttr): any relevant tensor attributes that correspond
+                to the old tensor. See :obj:`TensorAttr` documentation
+                for required and optional attributes. It is the job of
+                implementations of a FeatureStore to store this metadata in a
+                meaningful way that allows for tensor update from a
+                :obj:`TensorAttr` object.
+
+        Returns:
+            bool: whether the update was succesful.
+        """
+        attr = self._attr_cls.cast(attr)
+        self.remove_tensor(attr)
+        return self.put_tensor(tensor, attr)
+
+    # :obj:`AttrView` methods #################################################
+
+    def view(self, attr: Optional[TensorAttr]) -> AttrView:
+        r"""Returns an :obj:`AttrView` of the feature store, with the defined
+        attributes set."""
+        return AttrView(self, self._attr_cls.cast(attr))
+
+    # Python built-ins ########################################################
+
+    def __setitem__(self, key: TensorAttr, value: FeatureTensorType):
+        r"""Supports store[tensor_attr] = tensor."""
+        key = self._attr_cls.cast(key)
+        self.put_tensor(value, key)
+
+    def __getitem__(self, key: TensorAttr):
```

Comment:
`getattr` and `setattr` equivalents?

Reply:
I don't think a `FeatureStore` should implement `getattr` and `setattr`; imo, these should only be implemented for views on the store. I don't think it's particularly clean to have
```
store.group_name -> AttrView(group_name=group_name)
```
as this syntax seems more confusing to me than clarifying.

Reply:
Thought it makes sense to implement for stores without group names (like PyG `data` objects).

Reply:
Isn't the only way for an output to be fully specified through `getattr` to allow for chaining, which necessitates that `getattr` can return an `AttrView`? That feels odd to me.

Reply:
If `TensorAttr` sets both `group_name` and `index` to `None` by default, then `store.{attr_name}` should give you a tensor. I am okay with leaving this out for now.